### ISSUE TITLE
Category Mapping

### ISSUE THREAD

**Issue by skwp** [reverbdotcom/reverb-magento#20]
Category  Mapping

API: https://reverb.com/api/categories
Display: “name” property
In database: “slug” property for top level categories and for subcategories use “slug” and “product_type_slug”

Mapping table should lookl ike this:
magento_category_id: int
reverb_category_id:int

reverb_categories (synced on demand)
display_name:
reverb_product_type_slug: (top level categories call this “slug”)
reverb_category_slug: (subcategories call this ‘slug’ and also have a product_type_slug)

Screen looks like
Magento>Category>Tree  —- Dropdown of Reverb Category display names
Paginate so that it doesn’t load forever

**Comment by skwp** [comment]
Here's a screenshot of this implemented on BigCommerce which has a similar nested category structure

<img width="841" alt="screen shot 2015-09-09 at 2 03 18 pm" src="https://cloud.githubusercontent.com/assets/1731/9771401/9a7de1d6-56fb-11e5-8910-84a99d6b8430.png">

**Comment by skwp** [comment]
The way the mapping works is picking the most specific mapped category in the tree of categories. If magento allows for multiple categories  per  listing, just go with the first one that is mapped.

For example if we have
Guitar=>Electric=>Strings and only Electric is mapped, then when a product in "Strings" is synced, it will default to the Electric mapping

**Comment by dunagan5887** [comment]
1. So if a product has no category mapping, it's not going to synced to Reverb for listing creation or updates? Should the category only be included on creations, or also updates?
2. https://dev.reverb.com/docs/create-a-listing I'm not seeing any category fields here. Should the category be part of this API call? If so, what is the field and what should be the syntax? 
3. For the API call to fetch the categories, I'm assuming that it's just a GET to https://reverb.com/api/categories, correct? The Reverb system keeps that url updated with the most recent category set?

**Comment by skwp** [comment]
1. We need a switch I think because the current functionality is "always sync even without category". We need a switch that says "don't sync products without category mappings"
2. Updated docs, we have a "product_type" for the toplevel type and up to 2 additional "categories" which are subcategories. These are submitted as slugs and gotten from /api/categories

The Magento system should pull down the reverb categories only once. They change very infrequently. We may eventually provide a button to re-fetch them.

**Comment by dunagan5887** [comment]
2. Do you have a link to the Updated docs I can check out?

If Reverb categories should only be pulled down once, what' the purpose of the "Provide a button on the category screen that says "Resync categories" spec?

**Comment by skwp** [comment]
Yep just refresh this page https://dev.reverb.com/docs/create-a-listing

Yes, we should provide a button to resync categories, I'm just saying in terms of iterative implementation, first implement a single time sync and mapping, then we can think about the resync. I think there are some issues on resync such as figuring out what to keep/throw away that make it complicated. Use your discretion.

**Comment by dunagan5887** [comment]
In the interest of development time, would it make sense for me to just hard code that json object into the code for now? A onetime migration script would be created create the reverb categories from that json.

**Comment by skwp** [comment]
Totally fine with hardcoding the current json

**Comment by dunagan5887** [comment]
Sounds good

**Comment by skwp** [comment]
I would put it into a file probably. That way when we do bulid the sync button we can just dump to the same file. That's what we did on other platforms

**Comment by dunagan5887** [comment]
https://dev.reverb.com/docs/create-a-listing

This makes it seem like multiple reverb_category_slug values can be submitted in the "categories" field. Is this the desired functionality?

**Comment by skwp** [comment]
yes up to two categories can be submitted

**Comment by dunagan5887** [comment]
So will we need a patch to prevent more than 2 from being included?

**Comment by skwp** [comment]
I think the API will handle this gracefully and just take the first two
